# Data Cleaning

Data scientists spend a large amount of their time cleaning datasets and getting them down to a form with which they can work. In fact, a lot of data scientists argue that the initial steps of obtaining and cleaning data constitute 80% of the job.

## 1. Set up the environment and import the data

In [1]:
import pandas as pd

In [2]:
data = pd.read_csv('CAvideos.csv')

## 2. Investigate the data

In [3]:
# Take a peek at the first 2 rows
data.head(2)

video_id trending_date                                       title  \
0  n1WpP7iowLc      17.14.11  Eminem - Walk On Water (Audio) ft. Beyoncé   
1  0dBIkQ4Mz1M      17.14.11               PLUSH - Bad Unboxing Fan Mail   

  channel_title  category_id              publish_time  \
0    EminemVEVO           10  2017-11-10T17:00:03.000Z   
1     iDubbbzTV           23  2017-11-13T17:00:00.000Z   

                                                tags     views   likes  \
0  Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...  17158579  787425   
1  plush|"bad unboxing"|"unboxing"|"fan mail"|"id...   1014651  127794   

   dislikes  comment_count                                  thumbnail_link  \
0     43420         125882  https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg   
1      1688          13030  https://i.ytimg.com/vi/0dBIkQ4Mz1M/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  \
0              False             False                   False   
1              False             False                   False   

                                         description  
0  Eminem's new track Walk on Water ft. Beyoncé i...  
1  STill got a lot of packages. Probably will las...

In [4]:
# Look at the structure of the dataframe
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40881 entries, 0 to 40880
Data columns (total 16 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   video_id                40881 non-null  object
 1   trending_date           40881 non-null  object
 2   title                   40881 non-null  object
 3   channel_title           40881 non-null  object
 4   category_id             40881 non-null  int64 
 5   publish_time            40881 non-null  object
 6   tags                    40881 non-null  object
 7   views                   40881 non-null  int64 
 8   likes                   40881 non-null  int64 
 9   dislikes                40881 non-null  int64 
 10  comment_count           40881 non-null  int64 
 11  thumbnail_link          40881 non-null  object
 12  comments_disabled       40881 non-null  bool  
 13  ratings_disabled        40881 non-null  bool  
 14  video_error_or_removed  40881 non-null  bool  
 15  de

We see that we have 15 columns in all. However, currently,  some of them should change their data type. We also have missing values in the  “description” column. So, let’s clean up a little bit…

## 3. Clean up the data

### 3.1. Data types

First, we need to change the data type of some columns.
So we start with the "trending_date" column :

In [5]:
# Make lists of the columns which need attention and use this as reference to execute
to_change1 = [c for c in data.columns if "trending_date" in c] # numeric

# Change the target column data types
for c in to_change1:
    data[c] = data[c].apply(lambda x: pd.to_datetime(x, format = "%y.%d.%m"))

Then, we change the data type of the "publish_time" column :

In [6]:
# Make lists of the columns which need attention and use this as reference to execute
to_change2 = [c for c in data.columns if "publish_time" in c] # numeric

# Change the target column data types
for c in to_change2:
    data[c] = data[c].apply(lambda x: pd.to_datetime(x))

See the changes :

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40881 entries, 0 to 40880
Data columns (total 16 columns):
 #   Column                  Non-Null Count  Dtype              
---  ------                  --------------  -----              
 0   video_id                40881 non-null  object             
 1   trending_date           40881 non-null  datetime64[ns]     
 2   title                   40881 non-null  object             
 3   channel_title           40881 non-null  object             
 4   category_id             40881 non-null  int64              
 5   publish_time            40881 non-null  datetime64[ns, UTC]
 6   tags                    40881 non-null  object             
 7   views                   40881 non-null  int64              
 8   likes                   40881 non-null  int64              
 9   dislikes                40881 non-null  int64              
 10  comment_count           40881 non-null  int64              
 11  thumbnail_link          40881 non-null  o

### 3.2. Missing values

Now, let's see if our data had missing values :

In [8]:
# get the number of missing data points per column
data.isnull().sum()

video_id                     0
trending_date                0
title                        0
channel_title                0
category_id                  0
publish_time                 0
tags                         0
views                        0
likes                        0
dislikes                     0
comment_count                0
thumbnail_link               0
comments_disabled            0
ratings_disabled             0
video_error_or_removed       0
description               1296
dtype: int64

As we can see, the "description" column have 1296 missing values. So, we have to handle them by dropping the entire column (method 1), or by dropping the rows that have the missing values in that column (method 2), or by replacing the missig values (method 3). 

* Method 1 :

In [9]:
# remove all columns with at least one missing value
columns_with_na_dropped = data.dropna(axis=1)
columns_with_na_dropped.head(2)

video_id trending_date                                       title  \
0  n1WpP7iowLc    2017-11-14  Eminem - Walk On Water (Audio) ft. Beyoncé   
1  0dBIkQ4Mz1M    2017-11-14               PLUSH - Bad Unboxing Fan Mail   

  channel_title  category_id              publish_time  \
0    EminemVEVO           10 2017-11-10 17:00:03+00:00   
1     iDubbbzTV           23 2017-11-13 17:00:00+00:00   

                                                tags     views   likes  \
0  Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...  17158579  787425   
1  plush|"bad unboxing"|"unboxing"|"fan mail"|"id...   1014651  127794   

   dislikes  comment_count                                  thumbnail_link  \
0     43420         125882  https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg   
1      1688          13030  https://i.ytimg.com/vi/0dBIkQ4Mz1M/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  
0              False             False                   False  
1              False             False                   False

In [10]:
columns_with_na_dropped.isnull().sum()

video_id                  0
trending_date             0
title                     0
channel_title             0
category_id               0
publish_time              0
tags                      0
views                     0
likes                     0
dislikes                  0
comment_count             0
thumbnail_link            0
comments_disabled         0
ratings_disabled          0
video_error_or_removed    0
dtype: int64

* Method 2 :

In [11]:
# remove all rows with at least one missing value
rows_with_na_dropped = data.dropna(axis=0)
rows_with_na_dropped.head(2)

video_id trending_date                                       title  \
0  n1WpP7iowLc    2017-11-14  Eminem - Walk On Water (Audio) ft. Beyoncé   
1  0dBIkQ4Mz1M    2017-11-14               PLUSH - Bad Unboxing Fan Mail   

  channel_title  category_id              publish_time  \
0    EminemVEVO           10 2017-11-10 17:00:03+00:00   
1     iDubbbzTV           23 2017-11-13 17:00:00+00:00   

                                                tags     views   likes  \
0  Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...  17158579  787425   
1  plush|"bad unboxing"|"unboxing"|"fan mail"|"id...   1014651  127794   

   dislikes  comment_count                                  thumbnail_link  \
0     43420         125882  https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg   
1      1688          13030  https://i.ytimg.com/vi/0dBIkQ4Mz1M/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  \
0              False             False                   False   
1              False             False                   False   

                                         description  
0  Eminem's new track Walk on Water ft. Beyoncé i...  
1  STill got a lot of packages. Probably will las...

In [12]:
rows_with_na_dropped.isnull().sum()

video_id                  0
trending_date             0
title                     0
channel_title             0
category_id               0
publish_time              0
tags                      0
views                     0
likes                     0
dislikes                  0
comment_count             0
thumbnail_link            0
comments_disabled         0
ratings_disabled          0
video_error_or_removed    0
description               0
dtype: int64

* Method 3 :

In [13]:
# replace the missing values  
data.fillna("No description available", inplace=True)
data.isnull().sum()

video_id                  0
trending_date             0
title                     0
channel_title             0
category_id               0
publish_time              0
tags                      0
views                     0
likes                     0
dislikes                  0
comment_count             0
thumbnail_link            0
comments_disabled         0
ratings_disabled          0
video_error_or_removed    0
description               0
dtype: int64

Let's see our data now :

In [14]:
data.head()

video_id trending_date  \
0  n1WpP7iowLc    2017-11-14   
1  0dBIkQ4Mz1M    2017-11-14   
2  5qpjK5DgCt4    2017-11-14   
3  d380meD0W0M    2017-11-14   
4  2Vv-BfVoq4g    2017-11-14   

                                               title channel_title  \
0         Eminem - Walk On Water (Audio) ft. Beyoncé    EminemVEVO   
1                      PLUSH - Bad Unboxing Fan Mail     iDubbbzTV   
2  Racist Superman | Rudy Mancuso, King Bach & Le...  Rudy Mancuso   
3                           I Dare You: GOING BALD!?      nigahiga   
4        Ed Sheeran - Perfect (Official Music Video)    Ed Sheeran   

   category_id              publish_time  \
0           10 2017-11-10 17:00:03+00:00   
1           23 2017-11-13 17:00:00+00:00   
2           23 2017-11-12 19:05:24+00:00   
3           24 2017-11-12 18:01:41+00:00   
4           10 2017-11-09 11:04:14+00:00   

                                                tags     views    likes  \
0  Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...  17158579   787425   
1  plush|"bad unboxing"|"unboxing"|"fan mail"|"id...   1014651   127794   
2  racist superman|"rudy"|"mancuso"|"king"|"bach"...   3191434   146035   
3  ryan|"higa"|"higatv"|"nigahiga"|"i dare you"|"...   2095828   132239   
4  edsheeran|"ed sheeran"|"acoustic"|"live"|"cove...  33523622  1634130   

   dislikes  comment_count                                  thumbnail_link  \
0     43420         125882  https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg   
1      1688          13030  https://i.ytimg.com/vi/0dBIkQ4Mz1M/default.jpg   
2      5339           8181  https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg   
3      1989          17518  https://i.ytimg.com/vi/d380meD0W0M/default.jpg   
4     21082          85067  https://i.ytimg.com/vi/2Vv-BfVoq4g/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  \
0              False             False                   False   
1              False             False                   False   
2              False             False                   False   
3              False             False                   False   
4              False             False                   False   

                                         description  
0  Eminem's new track Walk on Water ft. Beyoncé i...  
1  STill got a lot of packages. Probably will las...  
2  WATCH MY PREVIOUS VIDEO ▶ \n\nSUBSCRIBE ► http...  
3  I know it's been a while since we did this sho...  
4  🎧: https://ad.gt/yt-perfect\n💰: https://atlant...

## 4. Save the final results back to a file

In [15]:
data.to_csv("CAvideos_Cleaning.xlsx")

Mission complete. Happy Coding!